File: process_captions.R
#!/usr/bin/env Rscript
# process_captions.R — Convert meeting transcript files to CSV with phase labels.
#
# Phase detection uses key phrases from the experimenter's speech.
# Transitions always alternate: "starts now" → "time up" → "starts now" → ...
# The first phase covers everything before the first trigger phrase.
#
# Usage:
#   Rscript process_captions.R transcript.txt
#   Rscript process_captions.R transcript.txt -o output.csv
#   Rscript process_captions.R transcript.txt --list-phases
#   Rscript process_captions.R transcript.txt --phases ideaGenerationOne,ideaGenerationTwo
#   Rscript process_captions.R transcript.txt \
#     --phase-names "introduction,ideaGenOne,break,ideaGenTwo,selection,debrief"

.libPaths(c("~/R/library", .libPaths()))
suppressPackageStartupMessages(library(optparse))

# ── Trigger patterns ──────────────────────────────────────────────────────────

PAT_STARTS <- paste(
  "your time starts now",
  "your [0-9]+[-\\s]minute timer starts now",
  sep = "|"
)

PAT_STOP <- paste(
  "your time is up",
  "(?<!\\w)time is up",
  "stop generating ideas",
  sep = "|"
)

# ── Default phase sequence ────────────────────────────────────────────────────
# "break" appears twice because there are two inter-session breaks.

DEFAULT_PHASES <- c(
  "introduction",
  "ideaGenerationOne",
  "break",
  "ideaGenerationTwo",
  "break",
  "ideaSelection",
  "debriefing"
)

# ── Build transition table ────────────────────────────────────────────────────
# Given N phase names, produce N-1 transitions.
# Odd-numbered transitions (1, 3, 5, ...) use PAT_STARTS.
# Even-numbered transitions (2, 4, 6, ...) use PAT_STOP.

build_transitions <- function(phases) {
  n <- length(phases)
  if (n < 2) stop("At least 2 phase names are required.")
  lapply(seq_len(n - 1), function(i) {
    list(
      pattern    = if (i %% 2 == 1) PAT_STARTS else PAT_STOP,
      next_phase = phases[i + 1]
    )
  })
}

# ── Parse transcript ──────────────────────────────────────────────────────────

parse_transcript <- function(lines) {
  header_pat <- "^\\[([^\\]]+)\\]\\s+(\\d{1,2}:\\d{2}:\\d{2})\\s*$"

  rows            <- list()
  current_speaker <- NULL
  current_time    <- NULL
  current_lines   <- character(0)

  flush_entry <- function() {
    if (!is.null(current_speaker) && length(current_lines) > 0) {
      text <- paste(trimws(current_lines[nzchar(trimws(current_lines))]),
                    collapse = " ")
      if (nzchar(text)) {
        rows[[length(rows) + 1]] <<- list(
          time    = current_time,
          speaker = current_speaker,
          content = text
        )
      }
    }
  }

  for (raw_line in lines) {
    line <- trimws(raw_line)
    m    <- regexec(header_pat, line, perl = TRUE)[[1]]
    if (m[1] != -1) {
      flush_entry()
      parts           <- regmatches(line, regexec(header_pat, line, perl = TRUE))[[1]]
      current_speaker <- trimws(parts[2])
      current_time    <- parts[3]
      current_lines   <- character(0)
    } else if (nzchar(line)) {
      current_lines <- c(current_lines, line)
    }
  }
  flush_entry()

  if (length(rows) == 0) return(data.frame())

  data.frame(
    time    = sapply(rows, `[[`, "time"),
    speaker = sapply(rows, `[[`, "speaker"),
    content = sapply(rows, `[[`, "content"),
    stringsAsFactors = FALSE
  )
}

# ── Assign phases ─────────────────────────────────────────────────────────────

assign_phases <- function(df, phases) {
  transitions   <- build_transitions(phases)
  pending       <- transitions
  current_phase <- phases[1]
  phase_col     <- character(nrow(df))

  for (i in seq_len(nrow(df))) {
    if (length(pending) > 0) {
      if (grepl(pending[[1]]$pattern, df$content[i],
                ignore.case = TRUE, perl = TRUE)) {
        current_phase <- pending[[1]]$next_phase
        pending       <- pending[-1]
      }
    }
    phase_col[i] <- current_phase
  }

  df$phase <- phase_col

  # Warn about phases never reached
  triggered  <- unique(phase_col)
  still_pending_phases <- unique(sapply(pending, `[[`, "next_phase"))
  truly_missing <- setdiff(still_pending_phases, triggered)
  if (length(truly_missing) > 0) {
    message("WARNING: Phase(s) never triggered in this file: ",
            paste(truly_missing, collapse = ", "))
  }

  df
}

# ── Phase summary ─────────────────────────────────────────────────────────────

print_phase_summary <- function(df, display_phases) {
  cat(sprintf("\n%-22s %5s  %8s  %8s\n", "Phase", "Rows", "First", "Last"))
  cat(strrep("-", 52), "\n")
  for (phase in display_phases) {
    sub_df <- df[df$phase == phase, ]
    n  <- nrow(sub_df)
    ft <- if (n > 0) sub_df$time[1] else "\u2014"
    lt <- if (n > 0) sub_df$time[n] else "\u2014"
    cat(sprintf("%-22s %5d  %8s  %8s\n", phase, n, ft, lt))
  }
  cat(strrep("-", 52), "\n")
  cat(sprintf("%-22s %5d\n\n", "TOTAL", nrow(df)))
}

# ── CLI ───────────────────────────────────────────────────────────────────────

option_list <- list(
  make_option(
    c("-o", "--output"),
    type    = "character",
    default = NULL,
    help    = "Output CSV path [default: input filename with .csv extension]",
    metavar = "FILE"
  ),
  make_option(
    "--phase-names",
    type    = "character",
    default = NULL,
    metavar = "NAMES",
    help    = paste0(
      "Ordered comma-separated phase names. ",
      "The first phase covers content before the first trigger. ",
      "Transitions alternate automatically: starts-now, time-up, starts-now, ... ",
      "Default: introduction,ideaGenerationOne,break,ideaGenerationTwo,break,ideaSelection,debriefing"
    )
  ),
  make_option(
    "--phases",
    type    = "character",
    default = NULL,
    metavar = "FILTER",
    help    = "Comma-separated phases to include in the output (all others are dropped)"
  ),
  make_option(
    "--list-phases",
    action  = "store_true",
    default = FALSE,
    help    = "Print a row count and timestamp summary per phase, then exit"
  )
)

parser <- OptionParser(
  usage       = "%prog [options] input_file",
  option_list = option_list,
  description = paste(
    "Convert a meeting transcript ([Speaker] HH:MM:SS format) to a CSV file",
    "with columns: time, speaker, phase, content.",
    "\n\nPhase transitions are detected from key phrases in the transcript:",
    "\n  Start trigger : 'your time starts now' / 'your N-minute timer starts now'",
    "\n  Stop  trigger : 'your time is up' / 'time is up' / 'stop generating ideas'"
  )
)

args <- parse_args(parser, positional_arguments = 1)
opts       <- args$options
input_file <- args$args[1]

if (!file.exists(input_file)) stop("File not found: ", input_file)

# Resolve phase list
if (!is.null(opts[["phase-names"]])) {
  phases <- trimws(strsplit(opts[["phase-names"]], ",")[[1]])
  if (length(phases) < 2) stop("--phase-names requires at least 2 phase names.")
} else {
  phases <- DEFAULT_PHASES
}

display_phases <- unique(phases)   # for summary / validation (deduped, ordered)

# Parse and assign
lines <- readLines(input_file, encoding = "UTF-8", warn = FALSE)
df    <- parse_transcript(lines)
if (nrow(df) == 0) stop("No transcript entries found in: ", input_file)

df <- assign_phases(df, phases)

# --list-phases
if (isTRUE(opts[["list-phases"]])) {
  print_phase_summary(df, display_phases)
  quit(status = 0)
}

# --phases filter
if (!is.null(opts$phases)) {
  requested <- trimws(strsplit(opts$phases, ",")[[1]])
  invalid   <- setdiff(requested, display_phases)
  if (length(invalid) > 0) {
    stop("Unknown phase(s): ", paste(invalid, collapse = ", "),
         "\nValid phases: ", paste(display_phases, collapse = ", "))
  }
  df <- df[df$phase %in% requested, ]
  if (nrow(df) == 0) stop("No rows remain after applying --phases filter.")
}

# Write output
output_file <- if (!is.null(opts$output)) {
  opts$output
} else {
  sub("\\.[^.]+$", ".csv", input_file)
}

df_out <- df[, c("time", "speaker", "phase", "content")]
write.csv(df_out, output_file, row.names = FALSE)

phases_present <- display_phases[display_phases %in% unique(df$phase)]
cat(sprintf("Done: %d entries written to %s\n", nrow(df_out), output_file))
cat(sprintf("      Phases included: %s\n", paste(phases_present, collapse = ", ")))

File: process_captions.py
#!/usr/bin/env python3
"""
process_captions.py — Convert closed caption files to CSV.

Supported formats:
  - Meeting transcript  [Speaker Name] HH:MM:SS / content  →  time, speaker, content
  - SRT (.srt)                                              →  index, start_time, end_time, duration_seconds, text
  - WebVTT (.vtt)                                           →  index, start_time, end_time, duration_seconds, text

Usage:
    python process_captions.py transcript.txt
    python process_captions.py transcript.txt -o output.csv
    python process_captions.py input.srt
    python process_captions.py input.vtt -o output.csv
    python process_captions.py input.srt --include-timestamps
"""

import re
import csv
import sys
import argparse
from pathlib import Path


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

def parse_time_seconds(time_str: str) -> float:
    """Convert a timestamp string to total seconds (float)."""
    time_str = time_str.strip().replace(",", ".")
    parts = time_str.split(":")
    if len(parts) == 3:
        hours, minutes, seconds = parts
        return int(hours) * 3600 + int(minutes) * 60 + float(seconds)
    elif len(parts) == 2:
        minutes, seconds = parts
        return int(minutes) * 60 + float(seconds)
    return float(time_str)


def format_time(seconds: float) -> str:
    """Format total seconds as HH:MM:SS.mmm string."""
    h = int(seconds // 3600)
    m = int((seconds % 3600) // 60)
    s = seconds % 60
    return f"{h:02d}:{m:02d}:{s:06.3f}"


def strip_tags(text: str) -> str:
    """Remove HTML/VTT inline tags."""
    return re.sub(r"<[^>]+>", "", text).strip()


# ---------------------------------------------------------------------------
# Meeting transcript parser  →  time, speaker, content
# ---------------------------------------------------------------------------

# Matches lines like:  [Speaker Name] 17:03:12
_TRANSCRIPT_HEADER = re.compile(r"^\[([^\]]+)\]\s+(\d{1,2}:\d{2}:\d{2})\s*$")


def is_transcript_format(content: str) -> bool:
    for line in content.splitlines():
        if _TRANSCRIPT_HEADER.match(line.strip()):
            return True
    return False


def parse_transcript(content: str) -> list[dict]:
    """
    Parse a meeting transcript where each entry is:

        [Speaker Name] HH:MM:SS
        Content line(s)…

    Consecutive content lines belonging to the same entry are joined with a
    space. Blank lines between entries are ignored.
    """
    rows = []
    current_speaker: str | None = None
    current_time: str | None = None
    current_lines: list[str] = []

    def flush():
        if current_speaker is not None and current_lines:
            text = " ".join(l.strip() for l in current_lines if l.strip())
            if text:
                rows.append({
                    "time": current_time,
                    "speaker": current_speaker,
                    "content": text,
                })

    for raw_line in content.splitlines():
        line = raw_line.strip()
        m = _TRANSCRIPT_HEADER.match(line)
        if m:
            flush()
            current_speaker = m.group(1).strip()
            current_time = m.group(2).strip()
            current_lines = []
        elif line == "":
            # blank line — ignore, don't flush yet (next header will flush)
            continue
        else:
            current_lines.append(line)

    flush()
    return rows


def write_transcript_csv(rows: list[dict], output_path: str) -> None:
    fields = ["time", "speaker", "content"]
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fields)
        writer.writeheader()
        writer.writerows(rows)


# ---------------------------------------------------------------------------
# SRT parser
# ---------------------------------------------------------------------------

def parse_srt(content: str) -> list[dict]:
    blocks = re.split(r"\n{2,}", content.strip())
    cues = []
    for block in blocks:
        lines = block.strip().splitlines()
        if len(lines) < 3:
            continue
        idx_line = lines[0].strip()
        index = idx_line if re.match(r"^\d+$", idx_line) else None
        time_line_idx = 1 if index is not None else 0

        time_line = lines[time_line_idx]
        time_match = re.match(
            r"(\d{1,2}:\d{2}:\d{2}[,\.]\d{1,3})\s*-->\s*(\d{1,2}:\d{2}:\d{2}[,\.]\d{1,3})",
            time_line,
        )
        if not time_match:
            continue

        start_s = parse_time_seconds(time_match.group(1))
        end_s = parse_time_seconds(time_match.group(2))
        text = strip_tags(" ".join(l.strip() for l in lines[time_line_idx + 1:] if l.strip()))

        cues.append({
            "index": index or str(len(cues) + 1),
            "start_time": format_time(start_s),
            "end_time": format_time(end_s),
            "start_seconds": round(start_s, 3),
            "end_seconds": round(end_s, 3),
            "duration_seconds": round(end_s - start_s, 3),
            "text": text,
        })
    return cues


# ---------------------------------------------------------------------------
# VTT parser
# ---------------------------------------------------------------------------

def parse_vtt(content: str) -> list[dict]:
    content = re.sub(r"^WEBVTT[^\n]*\n", "", content, flags=re.MULTILINE)
    for block_type in ("NOTE", "STYLE", "REGION"):
        content = re.sub(rf"{block_type}\s.*?(?=\n{{2,}}|\Z)", "", content, flags=re.DOTALL)

    cues = []
    for block in re.split(r"\n{2,}", content.strip()):
        lines = block.strip().splitlines()
        if not lines:
            continue
        time_line_idx = 0
        index = None
        if "-->" not in lines[0] and len(lines) > 1:
            index = lines[0].strip()
            time_line_idx = 1
        if time_line_idx >= len(lines):
            continue
        time_match = re.match(
            r"(\d{1,2}:\d{2}:\d{2}[,\.]\d{1,3}|\d{1,2}:\d{2}[,\.]\d{1,3})\s*-->\s*"
            r"(\d{1,2}:\d{2}:\d{2}[,\.]\d{1,3}|\d{1,2}:\d{2}[,\.]\d{1,3})",
            lines[time_line_idx],
        )
        if not time_match:
            continue
        start_s = parse_time_seconds(time_match.group(1))
        end_s = parse_time_seconds(time_match.group(2))
        text = strip_tags(" ".join(l.strip() for l in lines[time_line_idx + 1:] if l.strip()))

        cues.append({
            "index": index or str(len(cues) + 1),
            "start_time": format_time(start_s),
            "end_time": format_time(end_s),
            "start_seconds": round(start_s, 3),
            "end_seconds": round(end_s, 3),
            "duration_seconds": round(end_s - start_s, 3),
            "text": text,
        })
    return cues


def write_subtitle_csv(cues: list[dict], output_path: str, include_timestamps: bool) -> None:
    if include_timestamps:
        fields = ["index", "start_time", "end_time", "start_seconds", "end_seconds", "duration_seconds", "text"]
    else:
        fields = ["index", "start_time", "end_time", "duration_seconds", "text"]
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fields, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(cues)


# ---------------------------------------------------------------------------
# Format detection
# ---------------------------------------------------------------------------

def detect_format(content: str, path: Path) -> str:
    if content.lstrip().startswith("WEBVTT") or path.suffix.lower() == ".vtt":
        return "vtt"
    if is_transcript_format(content):
        return "transcript"
    return "srt"


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(
        description=(
            "Convert closed caption / meeting transcript files to CSV.\n\n"
            "Supports:\n"
            "  • Meeting transcripts  [Speaker] HH:MM:SS  →  time, speaker, content\n"
            "  • SRT (.srt)                               →  index, start_time, end_time, …\n"
            "  • WebVTT (.vtt)                            →  index, start_time, end_time, …"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("input", help="Input file (.txt transcript, .srt, or .vtt)")
    parser.add_argument(
        "-o", "--output",
        help="Output CSV path (default: same name as input with .csv extension)",
    )
    parser.add_argument(
        "--include-timestamps",
        action="store_true",
        help="(SRT/VTT only) Also include raw start_seconds and end_seconds columns",
    )
    args = parser.parse_args()

    input_path = Path(args.input)
    if not input_path.exists():
        print(f"Error: file not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    output_path = str(args.output or input_path.with_suffix(".csv"))
    content = input_path.read_text(encoding="utf-8-sig")
    fmt = detect_format(content, input_path)

    if fmt == "transcript":
        rows = parse_transcript(content)
        if not rows:
            print("Warning: no entries found in transcript.", file=sys.stderr)
            sys.exit(1)
        write_transcript_csv(rows, output_path)
        print(f"Done: {len(rows)} entries written to {output_path}  [columns: time, speaker, content]")

    elif fmt == "vtt":
        cues = parse_vtt(content)
        if not cues:
            print("Warning: no cues found in VTT file.", file=sys.stderr)
            sys.exit(1)
        write_subtitle_csv(cues, output_path, args.include_timestamps)
        print(f"Done: {len(cues)} cues written to {output_path}")

    else:
        cues = parse_srt(content)
        if not cues:
            print("Warning: no cues found in SRT file.", file=sys.stderr)
            sys.exit(1)
        write_subtitle_csv(cues, output_path, args.include_timestamps)
        print(f"Done: {len(cues)} cues written to {output_path}")


if __name__ == "__main__":
    main()
